Initial derivation remote cache related metrics: check, download and upload time.

restingbull opened the pull request:
* Introduces building block datum, LocalActions, to make it easier to reason about actions.
* Introduces RemoteCacheMetrics as basic reporting datum for cache performance.

saraadams commented:
@restingbull I haven't had a look at this, as I wanted to give you time to address Louis' comments first. Let me know if you are waiting for my review first.

restingbull commented:
Been a bit busy to address all the comments. Will get to it this week.

afrueda97 commented:
I've added the suggestions to the pull request @lrgoldstein , what do you think?

lrgoldstein commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/LocalActions.java:
```diff
@@ -0,0 +1,167 @@
+    if(isEmpty()) {
+      return "No local actions executed";
+    } else {
+      return "This is non-empty.";
```
Other providers just return `null`. Suggest being consistent.

lrgoldstein commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/LocalActionsDataProvider.java:
```diff
@@ -0,0 +1,81 @@
+        DatumSupplierSpecification.of(LocalActions.class, DatumSupplier.memoized(this::derive)));
+  }
+
+
```
Nit: too much space here.

restingbull commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetricsDataProvider.java:
```diff
@@ -0,0 +1,125 @@
+            * 100);
+  }
+
+  RemoteCacheData coalesce(LocalAction action) {
```
co·a·lesce
/ˌkōəˈles/
verb
come together to form one mass or whole.

I suppose we could use aggregate or summarize?

restingbull commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetricsDataProvider.java:
```diff
@@ -0,0 +1,125 @@
+
+  static class RemoteCacheData {
+
+    public final String name;
```
Future work, really. Was intending to add some actual useful data to the metrics.

restingbull commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java:
```diff
@@ -0,0 +1,84 @@
+
+  @Override
+  public String getEmptyReason() {
+    return "No Remote Cache checks or downloads";
```
That's a bit of an unexpected api.

lrgoldstein replied:
You would expect it to have an Empty Reason if it's not Empty? It's typically determined dynamically, as you could run into a number of different problems while producing a datum depending on the complexity of the calculation.

restingbull commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java:
```diff
@@ -0,0 +1,84 @@
+
+  @Override
+  public String getDescription() {
+    return "Remote cache usage.";
```
Added more description.

restingbull commented on analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Tests.

restingbull commented:
> > I've added the suggestions to the pull request @lrgoldstein , what do you think?
> 
> I've gone through and resolved the items that you addressed and added a couple... there were also a handful from before that weren't addressed (be sure to "Load" the "hidden conversations" in the middle).
> 
> Edit: Also, you have to [sign all your commits](https://github.com/EngFlow/bazel_invocation_analyzer/blob/main/CONTRIBUTING.md#process).

You have to do a bit more that just sign them. I have `commit.gpgsign true` (hence I have <img width="93" alt="Screen Shot 2023-03-01 at 8 29 34 PM" src="https://user-images.githubusercontent.com/58151731/222315556-7726cb3a-0c43-417e-868f-b42fb515da5e.png">) -- which means I can prove I made the commit...

You have to a more explicit `--sign-off`, which means you stand by it (I guess?). Thanks, git, for not making this confusing.
